Exit if STDIN is closed #1367

igoyak opened the pull request:
See https://github.com/ranger/ranger/issues/1367

I'm not sure if this is the preferred way to exit ranger, suggestions welcome

toonn commented on ranger/gui/ui.py:
```diff
@@ -272,6 +272,9 @@ def handle_input(self):
                 else:
                     if not self.fm.input_is_blocked():
                         self.handle_key(key)
+            elif key == -1 and not os.isatty(0):
```
Could you use `sys.stdin.fileno()` rather than the hard coded `0`?

igoyak replied:
Good idea, fixed.

toonn commented on ranger/gui/ui.py:
```diff
@@ -272,6 +272,9 @@ def handle_input(self):
                         self.handle_key(key)
+            elif key == -1 and not os.isatty(0):
+                # STDIN has been closed
+                sys.exit(1)
```
Proper way to exit is `self.fm.exit()`

igoyak replied:
Thanks, fixed.

toonn commented:
I'm pretty sure the other issues are the same problem. Feel free to include `Fixes #1367\nFixes #87...` in one of your commits.

igoyak commented:
Thanks for the feedback! I will check it out in a couple of days.

toonn commented:
Hmm, I actually can't reproduce the problem with the master version of ranger. @SuperShinyEyes, could you test this fix?

igoyak commented:
About propagating the SIGHUP: I've done some reading, but I'm not entirely sure how to do it, why it isn't happening currently or whether it's actually a good idea. It is related to the "job control" of shells and "process groups", where the "process leader" is responsible for forwarding signals (or rather translate a SIGHUP into SIGINTs for the children?). I could try to dig deeper, but maybe not in the foreseeable future. 

I do believe the behavior proposed in this PR is correct, though. Reasoning: ranger is a user-facing application. If the user cannot supply input, ranger has no reason to keep running.

igoyak commented:
> Hmm, I actually can't reproduce the problem with the master version of ranger. @SuperShinyEyes, could you test this fix?

Did you try the method using gdb? That is the only one that seems reliable. 
I rechecked both methods and can still produce the bug in commit 1ca1255a3cf26eb646fca75b5388b2c645fdf466

SuperShinyEyes commented:
@toonn @igoyak Is there a still need for me to test this fix? 

Sorry for the late reply. Last week was BUSY. I'd be more than happy to help fixing this. :)

SuperShinyEyes commented:
@toonn Got it. Will try tonight.

igoyak commented:
@toonn I have finally had more time to look deeper into this issue. 
I have updated the source branch of this PR, just rebased to use a modern commit as a parent. 
I have also written some more conclusions and a cross-platform way to replicate the issue: https://github.com/ranger/ranger/issues/1367#issuecomment-521773193

This issue still affects me (and others, it seems), and I think it would be a good idea to merge to master whenever appropriate.

toonn commented:
Will look into this tomorrow. Thanks for your effort!

@SuperShinyEyes, did you find the time to test this?

toonn commented:
Hmm, finally found time to look into this. Still can't reproduce by closing a terminal.
Apparently gdb won't run on macOS due to some code signing issues, at least without much effort. However, I was able to reproduce this with gdb on linux, had to guess at the type to cast the `close` call to though.

Now I can finally merge this 🎉 : )
